interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,353 @@
+			}
+			switch r2 {
+			case '*', '?', '[', ']', '{', '}', '\\':
+				runes = append(runes, r, r2)
+			default:
+				// do not add r to runes if it's unnecessary
+				runes = append(runes, r2)
+			}
+		case '[', ']', '{', '}':
+			// Should be impossible, we just rendered this variant
+			return nil, fmt.Errorf(`internal error: unexpected unescaped '%v' character`, r)
+		default:
+			runes = append(runes, r)
+		}
+	}
+
+	consumeText()
+
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator, compGlobstar}) {
+		// If components end with /**/*, strip trailing /*
+		components = components[:len(components)-2]
+	}
+
+	// Add terminal marker or convert existing doublestar to terminal doublestar
+	if prevComponentsAre([]componentType{compSeparatorDoublestar, compSeparator}) {
+		components = components[:len(components)-2]
+		components = append(components, component{compType: compSeparatorDoublestarSeparatorTerminal})
+	} else if prevComponentsAre([]componentType{compSeparatorDoublestar}) {
+		components[len(components)-1] = component{compType: compSeparatorDoublestarTerminal}
+	} else {
+		components = append(components, component{compType: compTerminal})
+	}
+
+	var variantBuf strings.Builder
+	var regexBuf strings.Builder
+	regexBuf.WriteRune('^')
+	for _, c := range components {
+		variantBuf.WriteString(c.String())
+		regexBuf.WriteString(c.componentRegex())
+	}
+	regexBuf.WriteRune('$')
+	regex := regexp.MustCompile(regexBuf.String())
+	regex.Longest()
+
+	v := PatternVariant{
+		variant:    variantBuf.String(),
+		components: components,
+		regex:      regex,
+	}
+
+	return &v, nil
+}
+
+// Need to replace unescaped "**", but must be careful about an escaped '\\'
+// before the first '*', since that doesn't escape the first '*'.
+var doublestarReplacer = regexp.MustCompile(`([^\\](\\\\)*)\*\*`)
+
+func prepareVariantForParsing(variant string) string {
+	prepared := doublestarReplacer.ReplaceAllStringFunc(variant, func(s string) string {
+		// Discard trailing "**", add "⁑" instead
+		return s[:len(s)-2] + "⁑"
+	})
+	return prepared
+}
+
+type componentReader struct {
+	components []component
+	submatches []string
+	index      int
+}
+
+func (r *componentReader) next() (*component, string) {
+	if r.index >= len(r.components) {
+		return &component{compType: compTerminal}, ""
+	}
+	comp := &r.components[r.index]
+	submatch := r.submatches[r.index]
+	r.index++
+	return comp, submatch
+}
+
+// Compare returns the relative precence of the receiver and the given pattern
+// variant when considering the given matching path.
+//
+// Returns one of the following, if no error occurs:
+// -1 if v has lower precedence than other
+// 0 if v and other have equal precedence (only possible if v == other)
+// 1 if v has higher precedence than other.
+func (v *PatternVariant) Compare(other *PatternVariant, matchingPath string) (int, error) {
+	selfSubmatches := v.regex.FindStringSubmatch(matchingPath)
```

Comment:
won't this and the potentially other FindStringSubmatch end up being calle a number of times for the same matchingPath when going over the matched variants list?

Reply:
Yes, it will. A very early implementation of this PR started with all given patterns, then progressively eliminated all but the highest-precedence patterns after looking at the next rune in each pattern. We certainly could use a similar approach here, where we instead look at the next component.

I chose to compare two variants at a time in order to make the precedence between two variants as clear as possible. If we use a filter-like approach, then there's some added complexity regarding keeping track of all of the current highest precedence patterns and throwing them out whenever a higher precedence pattern is seen. Also, we need to keep track of how far into the pattern we've looked (the $n^{th}$ component). This is all doable, I'll explore it tomorrow and see how it looks/feels.

Reply:
FWIW I think this can be a followup

Reply:
Let's do this in a follow-up. That way the precedence logic is clear enough now, and we can improve performance later while making it clear how the logic remains unchanged.